### DISCUSSION TITLE
Language detection testing

### DISCUSSION THREAD

**Discussion by ahnpnl** [robisim74/qwik-speak#72] [Q&A]
Hi,

I’m looking for a way to test translation texts to add to my pipeline. The idea is similar to e2e but in this case, the tests will grab all texts in the page to check if those texts are in given language.

I wonder if anyone has done anything like that?

**Comment by robisim74** [comment]
I don't know if I understand: you would just like to know if _all the texts_ on the page are in a certain language (i.e. have they been translated correctly)?

**Reply by ahnpnl** [reply]
Ah yes, that’s what I wanted to do

**Reply by robisim74** [reply]
Uhm... extracting all the texts on a rendered page is not so simple: furthermore some texts could be deliberately in another language.

Just an idea: add a` lang="{lang}"` attribute to each html tag with a text you want to extract, extract them from the html, and then use a language detection library or ChatGPT to determine the language?

**Reply by ahnpnl** [reply]
Hmm, that sounds quite some work to modify DOM in unit test. 

Do we have a way for example:

- first time rendering component with English translations. Store the HTML string in a `const`
- trigger something to load translations in Italian and store the new HTML string in a `const`.
- compare the 2 `const` to see if they are different.

It is the simplest way I usually do with SPA is like that. I just want to verify if the new translations are loaded correctly.

**Reply by robisim74** [reply]
This is different from checking that all texts are in a given language... But maybe I still haven't fully understood what you want to check!

However why a simple e2e test like [this](https://github.com/robisim74/qwik-speak/blob/main/src/e2e/home.spec.ts) is not enough for you?

**Reply by ahnpnl** [reply]
Hmm can we do that with unit test? I was also thinking a way is:
- render component with `QwikSpeak` initial locale is `en-US`. Capture the inner html string
- Render again the component but this time with initial locale `it-IT`. Capture the inner html string
- Now I can compare these 2 strings, I think this should work as minimum what I need.
